site/content/blog/2021-04-23-optimizing-go-compiler-1.md

```diff
@@ -0,0 +1,227 @@
+
+And that's exactly what the rewrite rules are for! You can take a look at [the specific syntax of the rules](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/gen/rulegen.go#L37-L60) but, in short, they contain:
+
+1.  A matching pattern, with an optional boolean condition preceded by `&&`
+2.  A right arrow `=>`
+3.  A new instruction that will override the matched pattern
+
+There are several rules to cover the conversion of multiplications by powers of two into left shifts in the [`generic.rules` file](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/gen/generic.rules#L172-L180). Let's take a look only at the first one, as all the others are identical in structure:
+
+```
+(Mul8  <t> n (Const8  [c])) && isPowerOfTwo8(c)
+=>
+(Lsh8x64  <t> n (Const64 <typ.UInt64> [log8(c)]))
+```
+
+We can see all the parts of a normal rule here:
+
+-   The matching pattern, `(Mul8 <t> n (Const8 [c]))`, which matches every multiplication instruction of a variable `n` by a constant `c`.
+-   The boolean condition, `isPowerOfTwo8(c)`, which restricts the pattern to those multiplications where the constant `c` is a power of two.
+-   The new instruction, `(Lsh8x64 <t> n (Const64 <typ.UInt64> [log8(c)]))`, which shifts `n` to the left `log(c)` places.
+
+It is ok if you do not understand every specific detail of the instructions, we will see them in detail a bit later in this post. Here, for example, we can note that the `8` suffixes make reference to the fact that this rule only matches multiplications of 8-bit variables. The other highlighted rules in the link to the `generic.rules` file cover the cases for variables of different bit widths.
+
+For now, it's enough to understand that the rules match a specific pattern of instructions, and rewrite those instructions with a new one that is hopefully more efficient. Also, it is important to know that there are generic rules, as the one we just studied, which are applied regardless of the specific architecture we are compiling for, and architecture-specific rules, such as the ones in the [AMD64.rules file](https://github.com/golang/go/blob/master/src/cmd/compile/internal/ssa/gen/AMD64.rules), which take advantage of the specifics of each architecture, further optimizing the code.
+
+## Connecting all together
+
+In order to investigate the issue and solve it, we will need to set up the development environment for the Go compiler. I encourage you to read the official [documentation](https://golang.org/doc/contribute): it is comprehensive and easy to understand. There are some particularities to the process, as the need to use Gerrit instead of Github, but it is really straightforward.
```

Comment:
It is not really necessary to use Gerrit. You can just use GitHub too. https://github.com/golang/go/wiki/GerritBot#workflow

Reply:
That's true, there's no "need" to use it, but [it is encouraged](https://golang.org/doc/contribute#sending_a_change_gerrit) (although using Github for newcomers is [also encouraged](https://golang.org/doc/contribute#sending_a_change_github)). I left the phrase without the explanation between the commas. Thanks!